File: frontend/src/services/transport.ts
import api from "./api";
import type {
  SaleSeatSelectionProductsResponse,
  SeatAdminAssignmentPayload,
  SeatAssignmentPayload,
  SeatBlockPayload,
  SeatHistoryResponse,
  SeatMapContext,
  SeatPostSignatureStatus,
  SeatSelectionContext,
  TripTransportConfigOut,
  TripTransportConfigPayload,
  VehicleLayoutDetail,
  VehicleLayoutListResponse,
  VehicleLayoutPayload,
  VehicleListResponse,
  VehicleOut,
  VehiclePayload,
} from "../types/transport";

export const getPostSignatureSeatStatus = (token: string) =>
  api.get<SeatPostSignatureStatus>(`/public/transport/signatures/${token}/status`);

export const getPostSaleSeatStatus = (token: string) =>
  api.get<SeatPostSignatureStatus>(`/public/transport/sales/${token}/status`);

export const getSaleSeatSelectionProducts = (token: string) =>
  api.get<SaleSeatSelectionProductsResponse>(`/public/transport/sales/${token}/products`);

export const getPublicSeatSelectionContext = (token: string, tripVehicleId?: number, departureId?: number) =>
  api.get<SeatSelectionContext>(`/public/transport/signatures/${token}/seats`, {
    params: {
      ...(tripVehicleId ? { trip_vehicle_id: tripVehicleId } : {}),
      ...(departureId ? { departure_id: departureId } : {}),
    },
  });

export const getPublicSaleSeatSelectionContext = (
  token: string,
  tripVehicleId?: number,
  departureId?: number,
  productPublicId?: string | null,
) =>
  api.get<SeatSelectionContext>(`/public/transport/sales/${token}/seats`, {
    params: {
      ...(tripVehicleId ? { trip_vehicle_id: tripVehicleId } : {}),
      ...(departureId ? { departure_id: departureId } : {}),
      ...(productPublicId ? { product_public_id: productPublicId } : {}),
    },
  });

export const selectSeatForSignature = (token: string, payload: SeatAssignmentPayload, departureId?: number) =>
  api.post<SeatSelectionContext>(`/public/transport/signatures/${token}/seats`, payload, {
    params: departureId ? { departure_id: departureId } : undefined,
  });

export const selectSeatForSale = (
  token: string,
  payload: SeatAssignmentPayload,
  departureId?: number,
  productPublicId?: string | null,
) =>
  api.post<SeatSelectionContext>(`/public/transport/sales/${token}/seats`, payload, {
    params: {
      ...(departureId ? { departure_id: departureId } : {}),
      ...(productPublicId ? { product_public_id: productPublicId } : {}),
    },
  });

export const listVehicleLayouts = () => api.get<VehicleLayoutListResponse>("/transport/layouts");

export const createVehicleLayoutRequest = (payload: VehicleLayoutPayload) =>
  api.post<VehicleLayoutDetail>("/transport/layouts", payload);

export const getVehicleLayoutDetail = (layoutId: number) =>
  api.get<VehicleLayoutDetail>(`/transport/layouts/${layoutId}`);

export const updateVehicleLayoutRequest = (layoutId: number, payload: VehicleLayoutPayload) =>
  api.put<VehicleLayoutDetail>(`/transport/layouts/${layoutId}`, payload);

export const duplicateVehicleLayout = (layoutId: number) =>
  api.post<VehicleLayoutDetail>(`/transport/layouts/${layoutId}/duplicate`);

export const deleteVehicleLayoutRequest = (layoutId: number) =>
  api.delete(`/transport/layouts/${layoutId}`);

export const listVehiclesRequest = () => api.get<VehicleListResponse>("/transport/vehicles");

export const createVehicleRequest = (payload: VehiclePayload) =>
  api.post<VehicleOut>("/transport/vehicles", payload);

export const updateVehicleRequest = (vehicleId: number, payload: VehiclePayload) =>
  api.put<VehicleOut>(`/transport/vehicles/${vehicleId}`, payload);

export const deleteVehicleRequest = (vehicleId: number) =>
  api.delete(`/transport/vehicles/${vehicleId}`);

export const getTripTransportConfig = (publicId: string) =>
  api.get<TripTransportConfigOut>(`/transport/products/${publicId}/transport-config`);

export const saveTripTransportConfig = (publicId: string, payload: TripTransportConfigPayload) =>
  api.put<TripTransportConfigOut>(`/transport/products/${publicId}/transport-config`, payload);

export const getSeatMapContext = (publicId: string, tripVehicleId?: number, departureId?: number) =>
  api.get<SeatMapContext>(`/transport/products/${publicId}/seat-map`, {
    params: {
      ...(tripVehicleId ? { trip_vehicle_id: tripVehicleId } : {}),
      ...(departureId ? { departure_id: departureId } : {}),
    },
  });

export const assignSeatAdmin = (
  publicId: string,
  payload: SeatAdminAssignmentPayload,
  tripVehicleId?: number,
  departureId?: number,
) =>
  api.post<SeatMapContext>(`/transport/products/${publicId}/seats/assignments`, payload, {
    params: {
      ...(tripVehicleId ? { trip_vehicle_id: tripVehicleId } : {}),
      ...(departureId ? { departure_id: departureId } : {}),
    },
  });

export const removePassengerSeat = (publicId: string, passengerId: number) =>
  api.delete(`/transport/products/${publicId}/seats/passengers/${passengerId}`);

export const blockSeatRequest = (
  publicId: string,
  payload: SeatBlockPayload,
  tripVehicleId?: number,
  departureId?: number,
) =>
  api.post<SeatMapContext>(`/transport/products/${publicId}/seats/block`, payload, {
    params: {
      ...(tripVehicleId ? { trip_vehicle_id: tripVehicleId } : {}),
      ...(departureId ? { departure_id: departureId } : {}),
    },
  });

export const getSeatHistory = (publicId: string, limit = 20, departureId?: number) =>
  api.get<SeatHistoryResponse>(`/transport/products/${publicId}/seat-history`, {
    params: {
      limit,
      ...(departureId ? { departure_id: departureId } : {}),
    },
  });

File: backend/app/api/v1/endpoints/transport.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps import get_current_active_user, get_db
from app.models.product import Product
from app.models.user import User
from app.models.transport import SeatAssignmentActor
from app.schemas.transport import (
    SeatAdminAssignmentPayload,
    SeatBlockPayload,
    SeatHistoryResponse,
    SeatMapContext,
    TripTransportConfigOut,
    TripTransportConfigPayload,
    VehicleLayoutDetail,
    VehicleLayoutListResponse,
    VehicleLayoutPayload,
    VehicleListResponse,
    VehicleOut,
    VehiclePayload,
)
from app.services.products import get_product_by_public_id
from app.services.transport import (
    assign_seat_to_passenger,
    block_trip_seat,
    create_vehicle,
    create_vehicle_layout,
    delete_vehicle,
    delete_vehicle_layout,
    duplicate_vehicle_layout,
    drop_passenger_assignment,
    get_trip_seat_map,
    get_trip_transport_config,
    get_vehicle_layout_detail,
    list_seat_history,
    list_vehicle_layouts,
    list_vehicles,
    save_trip_transport_config,
    update_vehicle,
    update_vehicle_layout,
)

router = APIRouter()


def _product(public_id: str, user: User, db: Session) -> Product:
    return get_product_by_public_id(public_id, user, db)


@router.get("/layouts", response_model=VehicleLayoutListResponse)
def list_layouts_endpoint(user: User = Depends(get_current_active_user), db: Session = Depends(get_db)) -> VehicleLayoutListResponse:
    return list_vehicle_layouts(user, db)


@router.post("/layouts", response_model=VehicleLayoutDetail)
def create_layout_endpoint(
    payload: VehicleLayoutPayload,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> VehicleLayoutDetail:
    return create_vehicle_layout(user, payload, db)


@router.get("/layouts/{layout_id}", response_model=VehicleLayoutDetail)
def get_layout_endpoint(
    layout_id: int,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> VehicleLayoutDetail:
    return get_vehicle_layout_detail(user, layout_id, db)


@router.put("/layouts/{layout_id}", response_model=VehicleLayoutDetail)
def update_layout_endpoint(
    layout_id: int,
    payload: VehicleLayoutPayload,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> VehicleLayoutDetail:
    return update_vehicle_layout(user, layout_id, payload, db)


@router.post("/layouts/{layout_id}/duplicate", response_model=VehicleLayoutDetail)
def duplicate_layout_endpoint(
    layout_id: int,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> VehicleLayoutDetail:
    return duplicate_vehicle_layout(user, layout_id, db)


@router.delete("/layouts/{layout_id}", status_code=204)
def delete_layout_endpoint(
    layout_id: int,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> None:
    delete_vehicle_layout(user, layout_id, db)


@router.get("/vehicles", response_model=VehicleListResponse)
def list_vehicles_endpoint(
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> VehicleListResponse:
    return list_vehicles(user, db)


@router.post("/vehicles", response_model=VehicleOut)
def create_vehicle_endpoint(
    payload: VehiclePayload,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> VehicleOut:
    return create_vehicle(user, payload, db)


@router.put("/vehicles/{vehicle_id}", response_model=VehicleOut)
def update_vehicle_endpoint(
    vehicle_id: int,
    payload: VehiclePayload,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> VehicleOut:
    return update_vehicle(user, vehicle_id, payload, db)


@router.delete("/vehicles/{vehicle_id}", status_code=204)
def delete_vehicle_endpoint(
    vehicle_id: int,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> None:
    delete_vehicle(user, vehicle_id, db)


@router.get("/products/{public_id}/transport-config", response_model=TripTransportConfigOut)
def get_transport_config_endpoint(
    public_id: str,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> TripTransportConfigOut:
    product = _product(public_id, user, db)
    return get_trip_transport_config(product, db)


@router.put("/products/{public_id}/transport-config", response_model=TripTransportConfigOut)
def save_transport_config_endpoint(
    public_id: str,
    payload: TripTransportConfigPayload,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> TripTransportConfigOut:
    product = _product(public_id, user, db)
    return save_trip_transport_config(product, payload, db)


@router.get("/products/{public_id}/seat-map", response_model=SeatMapContext)
def get_seat_map_endpoint(
    public_id: str,
    trip_vehicle_id: int | None = None,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> SeatMapContext:
    product = _product(public_id, user, db)
    return get_trip_seat_map(product, db, trip_vehicle_id=trip_vehicle_id)


@router.post("/products/{public_id}/seats/assignments", response_model=SeatMapContext)
def assign_seat_endpoint(
    public_id: str,
    payload: SeatAdminAssignmentPayload,
    trip_vehicle_id: int | None = None,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> SeatMapContext:
    product = _product(public_id, user, db)
    assign_seat_to_passenger(
        product,
        payload,
        actor=SeatAssignmentActor.admin,
        db=db,
        user=user,
        notes=payload.notes,
        status_override=payload.assignment_status,
    )
    return get_trip_seat_map(product, db, trip_vehicle_id=trip_vehicle_id)


@router.delete("/products/{public_id}/seats/passengers/{passenger_id}", status_code=204)
def remove_assignment_endpoint(
    public_id: str,
    passenger_id: int,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> None:
    product = _product(public_id, user, db)
    drop_passenger_assignment(
        product,
        passenger_id=passenger_id,
        actor=SeatAssignmentActor.admin,
        db=db,
        user=user,
        reason="Removido manualmente",
    )


@router.post("/products/{public_id}/seats/block", response_model=SeatMapContext)
def block_seat_endpoint(
    public_id: str,
    payload: SeatBlockPayload,
    trip_vehicle_id: int | None = None,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> SeatMapContext:
    product = _product(public_id, user, db)
    block_trip_seat(product, payload, db, user=user)
    return get_trip_seat_map(product, db, trip_vehicle_id=trip_vehicle_id)


@router.get("/products/{public_id}/seat-history", response_model=SeatHistoryResponse)
def seat_history_endpoint(
    public_id: str,
    limit: int = 20,
    user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> SeatHistoryResponse:
    product = _product(public_id, user, db)
    normalized_limit = max(1, min(limit, 100))
    return list_seat_history(product, normalized_limit, db)
